optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/index/ExternalizedIndexVariableSupply.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.domain.variable.index;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import org.optaplanner.core.api.score.director.ScoreDirector;
+import org.optaplanner.core.impl.domain.entity.descriptor.EntityDescriptor;
+import org.optaplanner.core.impl.domain.variable.descriptor.ListVariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.descriptor.VariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.listener.SourcedVariableListener;
+
+/**
+ * Alternative to {@link IndexVariableListener}.
+ */
+public class ExternalizedIndexVariableSupply<Solution_>
+        implements SourcedVariableListener<Solution_, Object>, IndexVariableSupply {
+
+    protected final ListVariableDescriptor<Solution_> sourceVariableDescriptor;
+
+    protected Map<Object, Integer> indexMap = null;
+
+    public ExternalizedIndexVariableSupply(ListVariableDescriptor<Solution_> sourceVariableDescriptor) {
+        this.sourceVariableDescriptor = sourceVariableDescriptor;
+    }
+
+    @Override
+    public VariableDescriptor<Solution_> getSourceVariableDescriptor() {
+        return sourceVariableDescriptor;
+    }
+
+    @Override
+    public void resetWorkingSolution(ScoreDirector<Solution_> scoreDirector) {
+        EntityDescriptor<Solution_> entityDescriptor = sourceVariableDescriptor.getEntityDescriptor();
+        List<Object> entityList = entityDescriptor.extractEntities(scoreDirector.getWorkingSolution());
+        indexMap = new HashMap<>();
```

Comment:
(important) Use an new IdentityHashMap() instead of a HashMap to avoid being affected by how the user decides to define equals/hashcode on their domain objects.

Reply:
Absolutely! I overlooked the fact that although the map holds _Integers_, user domain objects are the keys!